Add useStoreSelectors hook

adamziel opened the pull request:
## Description
This PR evolved into a shorthand hook called `useStoreSelectors`, used like this:

```js
const {
	selectionStart,
	selectionEnd,
} = useStoreSelectors(
         'core/editor',
	( { getEditorSelectionStart, getEditorSelectionEnd } ) => ( {
		selectionStart: getEditorSelectionStart(),
		selectionEnd: getEditorSelectionEnd(),
	} )
);
```

## Dev notes

### Added useStoreSelectors hook

`useStoreSelectors` is a custom react hook for retrieving selectors from registered stores. It could be considered a shorthand for `useSelect()` that provides all the selectors from a given store. For example:

```js
const { useStoreSelectors } = wp.data;

function HammerPriceDisplay( { currency } ) {
  const price = useStoreSelectors(
    'my-shop',
    ( { getPrice } ) => getPrice( 'hammer', currency ),
    [ currency ]
  );
  return new Intl.NumberFormat( 'en-US', {
    style: 'currency',
    currency,
  } ).format( price );
}

// Rendered in the application:
// <HammerPriceDisplay currency="USD" />
```

In the above example, when `HammerPriceDisplay` is rendered into an application, the price will be retrieved from the store state using the `mapSelectors` callback on `useStoreSelectors`. If the currency prop changes then any price in the state for that currency is retrieved. If the currency prop doesn't change and other props are passed in that do change, the price will
not change because the dependency is just the currency.


## Original description for posterity:

Every key stroke in the editor re-runs all registered selectors, sometimes multiple times. `useSelect()` is great for monitoring state changes. but problem with it is that the `mapSelect` function is executed on each registry change, related or not. Fortunately the component is only re-rendered if the newly computed value is different from the last one, but the selector is still executed every time.

This PR proposes a first argument to `useSelect` called `store`. If provided, `useSelect` will only subscribe to specific stores instead of the entire registry. Consider the following setup:

```jsx
const MyComponent = () => {
    const someStatistic = useSelect(( select ) => {
        console.log("Value re-computed");
        return select('core/edit-widgets').getSomeStatistic()
    });
    console.log("Component re-rendered");
    const { editEntityRecord } = useDispatch( 'core' );
    return (
        <div>
            Some statistic: {someStatistic}
            <textarea onInput=((e) => { editEntityRecord('root', 'postType', 15, { value: e.target.value } }) />
        </div>
    );
}
```

Typing "Gutenberg" in the textarea would log the following:

**Before:**
```
console.log("Value re-computed"); // Initial render
console.log("Component re-rendered");
console.log("Value re-computed"); // G
console.log("Value re-computed"); // u
console.log("Value re-computed"); // t
console.log("Value re-computed"); // e
console.log("Value re-computed"); // n
console.log("Value re-computed"); // b
console.log("Value re-computed"); // e
console.log("Value re-computed"); // r
console.log("Value re-computed"); // g
```

**After:**
Adding store dependencies like:
```
    const someStatistic = useSelect( 'core/edit-widgets', ( { getSomeStatistic } ) => {
        console.log("Value re-computed");
        return getSomeStatistic();
    }, []);
```

Reduces the number of times the selector is triggered:
```
console.log("Value re-computed"); // Initial render
console.log("Component re-rendered");
```

Note that the following code implies that `deps` will be passed to `useCallback`: 
```jsx
useSelect( 'core/data', fn );
```
I think it's an okay limitation though as the alternative would be fairly complex to implement. One is better off using `deps` anyway.

## Other considerations

Specifying dependent stores for all selectors should yield noticeable performance gains, but we can go even further. Some ideas:

* Don't re-run `mapSelect` immediately, instead use throttle of let's say 10ms (https://github.com/WordPress/gutenberg/pull/26695)
* Replace dirty-checking with observables as much as possible
* Specify specific state branch as a dependency (entity type? entity id?)

## How has this been tested?
1. Confirm the unit tests pass
2. A follow-up PR will add a few usage to the widgets editor, it will be much easier to reason about the behavior then

## Types of changes
New feature (non-breaking change which adds functionality)

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. <!-- Check code: `npm run lint`, Guidelines: https://developer.wordpress.org/coding-standards/wordpress-coding-standards/javascript/ -->
- [x] My code follows the accessibility standards. <!-- Guidelines: https://developer.wordpress.org/coding-standards/wordpress-coding-standards/accessibility/ -->
- [x] My code has proper inline documentation. <!-- Guidelines: https://developer.wordpress.org/coding-standards/inline-documentation-standards/javascript/ -->
- [x] I've included developer documentation if appropriate. <!-- Handbook: https://developer.wordpress.org/block-editor/ -->
- [x] I've updated all React Native files affected by any refactorings/renamings in this PR. <!-- React Native mobile Gutenberg guidelines: https://github.com/WordPress/gutenberg/blob/master/docs/contributors/native-mobile.md -->

adamziel committed:
Allow useSelect per store

ellatrix commented:
What if we limit useSelect to a single store per call? If you want to use multiple store, make multiple calls. This is already the case for `useDispatch`.

```js
useSelect( 'core/block-editor', ( { getSomething } ) => return getSomething( dependency ), [ dependency ] );
```

That also makes it more similar to `useDispatch`, just with mapping included.

```js
useSelect( storeKey, mapping, dependencies );
useDispatch( storeKey );
```

adamziel commented:
@ellatrix I like it! let's try that

adamziel committed:
New signature: useSelect( 'store', mapSelect, deps )

adamziel commented on packages/data/src/components/use-select/index.js:
```diff
@@ -78,9 +80,22 @@ const renderQueue = createQueue();
-export default function useSelect( _mapSelect, deps ) {
-	const mapSelect = useCallback( _mapSelect, deps );
+export default function useSelect( storeKey, _mapSelect, deps ) {
+	if ( arguments.length === 3 ) {
+		[ storeKey, _mapSelect, deps ] = arguments;
+	} else {
+		// Backwards compatible signature
+		storeKey = null;
+		[ _mapSelect, deps ] = arguments;
+	}
```
This is here to keep the arguments in README and make the linter happy. I don't like it too much. The alternative is having a new `useStoreSelect` selector and deprecate this one, but I don't like a longer name for the recommended default. Thoughts?

ellatrix replied:
Backwards compatibility code is good to have?

`useStoreSelect` would be interesting if we export it for the modules creating the store instead of importing the general `useSelect` from the data module. Tbh, I think that makes more sense than hardcoding the keys as strings. @gziolo was also working on a PR to export keys from the components creating the stores.

adamziel replied:
It is good to have, I just don't like processing `arguments` like that.

> useStoreSelect would be interesting if we export it for the modules creating the store instead of importing the general useSelect from the data module. Tbh, I think that makes more sense than hardcoding the keys as strings. @gziolo was also working on a PR to export keys from the components creating the stores.

Sounds like same goes for `useDispatch()`. It would work, we'd just have to update all the store-registering modules and remember about exporting this in the newly created ones 🤔

adamziel replied:
Hm..

```js
// Calls the callback with a `select` function
import { useSelect } from '@wordpress/data';

// Calls the callback with module-specific selectors:
import { useSelect as useCoreDataSelect } from '@wordpress/core-data';
import { useSelect as useEditWidgetsSelect } from '@wordpress/edit-widgets';
```

I'm not sure if I like it or not :D

adamziel replied:
```js
// Deprecated, a function expects a module as an argument
import { useDispatch } from '@wordpress/data';

// Deprecated, a function that takes no arguments
import { useDispatch as useCoreDataDispatch } from '@wordpress/core-data';
import { useDispatch as useEditWidgetsDispatch } from '@wordpress/edit-widgets';
```

adamziel replied:
I'm intrigued though, I'll give it a try in another PR tomorrow.

adamziel commented:
@ellatrix I updated the code and PR description.

adamziel committed:
Hopefully fix mobile tests

ellatrix commented:
It would be nice it update e.g. all `useSelect` calls in the block editor package and maybe paragraph block and see how it impacts performance.

noisysocks commented:
Performance improvements aside, I prefer this API. There's less boilerplate and it makes `useSelect` and `useDispatch` consistent. What do you think about deprecating the old signature?

kevin940726 commented on packages/data/src/components/use-select/index.js:
```diff
@@ -78,9 +80,23 @@ const renderQueue = createQueue();
-export default function useSelect( _mapSelect, deps ) {
-	const mapSelect = useCallback( _mapSelect, deps );
+export default function useSelect( storeKey, _mapSelect, deps ) {
+	if ( arguments.length < 3 ) {
+		// Backwards compatible signature
+		[ storeKey, _mapSelect, deps ] = [
+			null,
+			arguments[ 0 ],
+			arguments[ 1 ],
+		];
+	}
```
Can we instead just check the type of the first argument? If it's a function then it's `mapSelect`, if it's a string then it's `storeKey`?

gziolo replied:
I agree that it might be handled differently.

In general, extending API surface for `useSelect` creates some challenges:
- how to update JSDoc comment to cover new usage to ensure that proper documentation is generated for developers
- how to announce new API in a way that it doesn't become confusing, is it clear when to use which form?

kevin940726 commented on packages/data/src/components/use-select/test/index.js:
```diff
@@ -203,4 +203,79 @@ describe( 'useSelect', () => {
 		);
 	} );
+
+	it( 'uses memoized selector if dependent stores do not change', async () => {
```
Nitpick: we don't need the `async` keyword here?

kevin940726 commented on packages/data/src/components/use-select/test/index.js:
```diff
@@ -203,4 +203,79 @@ describe( 'useSelect', () => {
+		registry.registerStore( 'store-1', storeConfig );
+		registry.registerStore( 'store-2', storeConfig );
+
+		const selectSpyFoo = jest
+			.fn()
+			.mockImplementation( ( { getCounter } ) => getCounter() );
```
Nitpick: We can just write it like

```suggestion
		const selectSpyFoo = jest.fn( ( { getCounter } ) => getCounter() );
```

gziolo replied:
Isn't it how Prettier formats it? I would personally leave all formatting choices to tools.

adamziel replied:
I think @kevin940726  meant omitting the `.mockImplementation` part

kevin940726 commented on packages/data/src/components/use-select/test/index.js:
```diff
@@ -203,4 +203,79 @@ describe( 'useSelect', () => {
+		} );
+		let renderer;
+		act( () => {
+			renderer = TestRenderer.create(
```
Nitpick: I'd prefer to just use RTL, but this works too.

kevin940726 replied:
Yeah, I just found out that other files here use `react-test-renderer` as well. I agree to use `react-test-renderer` as is for simplicity.

kevin940726 replied:
> RTL uses react-test-renderer as well.

For the most part, I think RTL mainly just uses `ReactDOM.render` but not `react-test-renderer`, but I got your point 😂

adamziel commented:
More explorations here https://github.com/WordPress/gutenberg/pull/26723

gziolo reviewed (commented):
Nice exploration. I wanted to point to prior work done in the same area at the tame when React hooks weren't so widely used, but some of the challenges might still be valid:
- #12877
- #13177

The main reason why this idea was abandoned is:

> There are good ideas here, but with consideration of registry-enhanced selectors as introduced by #13662, the ability to determine store dependencies from a withSelect has become much more complex than this naive implementation can support.

where #13662 allows to:

> In some situations, you want to build selectors that target multiple stores at the same time. Until now we were relying on the global select function but the issue is that it only targets the default registry. If we have a separate provider, this might not work as expected. In this PR, I'm introducing a `createRegistrySelector` helper used to mark a selector a cross-stores selector and providing a registry object.

An example of the selector that depends on another store:

https://github.com/WordPress/gutenberg/blob/a63cc627fa9de3ba9cfc3fcdf5ac07e0549dd166/packages/block-directory/src/store/selectors.js#L67-L76

A follow-up comment might be still relevant in this context: https://github.com/WordPress/gutenberg/pull/13177#issuecomment-465277070.

There is also this issue that should be probably addressed first, which is how to ensure that there is a simple way to identify a given store as in many aspects using hardcoded strings doesn't scale well. See my proposal in #26655.

gziolo commented:
There is one more thing that we should keep in mind when working towards this proposal. In some cases `useSelect` can use selectors from several stores and some data is only used to pass as arguments to other selectors or to create conditions to use with data returned from the hook. Some examples:

https://github.com/WordPress/gutenberg/blob/4292a04c065c73581ec73af5540b49dcabd98b72/packages/edit-post/src/components/header/fullscreen-mode-close/index.js#L16-L37

https://github.com/WordPress/gutenberg/blob/4292a04c065c73581ec73af5540b49dcabd98b72/packages/edit-post/src/components/header/header-toolbar/index.js#L40-L72

https://github.com/WordPress/gutenberg/blob/4292a04c065c73581ec73af5540b49dcabd98b72/packages/edit-post/src/components/sidebar/settings-sidebar/index.js#L32-L55

The more high-level components, the higher the chance that this optimization is going to be harder to apply.

nerrad commented:
My feedback here aligns very much along the lines of what @gziolo has already outlined. However, I do like the idea of  optionally being able to subscribe to a single store (this has come up numerous times).

Before committing to this significant change though, have you done any performance measurements to see what actual impact this has? Relying on `console.log` dumps is a signal of a _potential_ performance improvement but can also be misleading in a React context where actual impact is minimal. Would the performance checks via `npm run test-performance` measure these changes? If so it'd be nice to get some numbers to compare.

adamziel committed:
Propose useStoreSelectors

adamziel committed:
Revert README

adamziel committed:
Add comment

adamziel commented:
All good notes, thank you @gziolo and @nerrad !

> Relying on console.log dumps is a signal of a potential performance improvement but can also be misleading in a React context where actual impact is minimal.

Of course! My point was more illustrative than qualitative. I didn't talk about actual performance improvements because I  wasn't quite sure how to measure the potential gains without actually refactoring everything to store-specific selectors. Some baseline measures I took in chrome devtools showed that runSelector takes anywhere between 8%-13% of total blocking time depending on how I use the editor (fast typing, pressing enter, triggering toolbar). If I interact with the editor heavily and JS gets blocked occasionally, that's more like 8%-13% of the real time, e.g. 130ms out of every second is spent on just re-running selectors. I hoped maybe we could get that to ~20ms by just not calling most of them.

I spent more time thinking about this and I found a way to assess potential gains. I figured I could keep track of the stores used by each selector like this:

```js
const dependentStore = useRef();
const mapSelect = useCallback( ( select, registry ) => {
	registry.__experimentalStartRecordingSelectStores();
	const retval = _mapSelect( select( storeKey ), registry );
	const stores = registry.__experimentalStopRecordingSelectStores();
	const newDependentStore = stores.length === 1 ? stores[0] : false;
	if ( newDependentStore !== dependentStore.current ) {
		dependentStore.current = newDependentStore;
		forceRender();
	}
	return retval;
}, deps );
```

Which then makes it trivial to subscribe to either entire registry or just a single store, depending on the value of `stores`:

```js
	if ( dependentStore.current ) {
		unsubscribe = registry.stores[ dependentStore.current ].subscribe( onChange );
	} else {
		unsubscribe = registry.subscribe( onChange );
	}
```

Which, by the way, makes it possible to keep track of the exact stores a selector needs to subscribe to even when registry selectors are involved (cc @gziolo @kevin940726). Unfortunately, even with that in place I saw the same 8%-13% range in the performance monitor. Boo hoo. 

If that makes sense as a methodology, I think it concludes the part of the discussion about performance gains.

There's also the part about having a potentially nicer API like this:

```js
const {
	selectionStart,
	selectionEnd,
} = useStoreSelectors(
         'core/editor',
	( { getEditorSelectionStart, getEditorSelectionEnd } ) => ( {
		selectionStart: getEditorSelectionStart(),
		selectionEnd: getEditorSelectionEnd(),
	} )
);
```

Or even this, using a shorthand notation:

```js
const {
	selectionStart,
	selectionEnd,
} = useStoreSelectors( 'core/editor', [ 'selectionStart', 'selectionEnd' ] );
```

I updated this PR with a draft of what could become `useStoreSelectors`.

gziolo commented:
I guess it's very difficult to verify the overall impact on performance without doing a major refactor.

> There's also the part about having a potentially nicer API like this:
> 
> ```js
> const {
> 	selectionStart,
> 	selectionEnd,
> } = useStoreSelectors(
>          'core/editor',
> 	( { getEditorSelectionStart, getEditorSelectionEnd } ) => ( {
> 		selectionStart: getEditorSelectionStart(),
> 		selectionEnd: getEditorSelectionEnd(),
> 	} )
> );
> ```

One interesting thing about introducing `useStoreSelectors` is that we could revisit your explorations once this new API is more widely used. I like the new hook proposed.
 
> Or even this, using a shorthand notation:
> 
> ```js
> const {
> 	selectionStart,
> 	selectionEnd,
> } = useStoreSelectors( 'core/editor', [ 'selectionStart', 'selectionEnd' ] );
> ```
> 
> I updated this PR with a draft of what could become `useStoreSelectors`.

I'm not entirely sure how this shorthand version would work. Given the efforts required to explain how this notation should be used and in which cases it is applicable, it's probably better to remove it from the scope of this PR.

adamziel committed:
Remove shorthand notation and add unit tests

adamziel committed:
Update README

adamziel commented:
@gziolo Done, I removed the array shorthand and added tests to this PR. I may explore another selector just for the shorthand in another PR, but let's call it out of scope for this one.

adamziel committed:
Remove now unrelated change

adamziel committed:
Updated docstring

adamziel committed:
Updated tests

adamziel committed:
Use renderHook from react testing library

gziolo commented on packages/data/README.md:
```diff
@@ -712,6 +712,24 @@ _Returns_
 
+<a name="useStoreSelectors" href="#useStoreSelectors">#</a> **useStoreSelectors**
+
+A shorthand for useSelect() that returns all the selectors from a given store.
```
`useSelect` is explained as:

> Custom react hook for retrieving props from registered selectors.

Documentation should follow the same theme. Doesn't it rather returns data processed by selectors from a single store.

gziolo replied:
A usage example would be very helpful as well.

One more, why did you pick the ending `Selectors` rather the `Select` that the original hook uses?

adamziel replied:
> One more, why did you pick the ending Selectors rather the Select that the original hook uses?

The original goes like this: `useSelect( (select) => select('store').selector() )` - what you get is the `select` function, hence the name `useSelect` makes sense.

This hook goes like: `useStoreSelectors( ({selector}) => selector() )` - what you get are all the selectors from a given store, hence the name `useStoreSelectors` makes more sense than `useStoreSelect`.

gziolo replied:
Yes, it sounds convincing. 👍 

Aside: I'm wondering if @nerrad had the same mindset when proposing `useSelect` :)

nerrad replied:
As a matter of fact: #15512

gziolo reviewed (commented):
I'm personally happy with the proposal. It will allow simplifying the usage of hooks in the lower level packages that use a very limited number of stores. In addition, it's very convenient for plugin authors that are concerned only about their custom stores.

adamziel committed:
Add usage example

adamziel committed:
Updated doc

adamziel committed:
Update CHANGELOG.md

gziolo commented:
One more thing that would be useful for those using `@wordpress/data` in other projects, an entry in CHANGELOG file similar to one added for `useSelect`: 
https://github.com/WordPress/gutenberg/blob/master/packages/data/CHANGELOG.md#new-feature

youknowriad commented:
Hey there! I'm not really convinced about this new hook, for me it's just a duplicate API over useSelect. it's a bit nicer for sure if you just need selectors from a store but that's rather a specific use-case and I'm not sure we want to maintain duplicate APIs for this.

youknowriad commented:
Also, why selectors and not actions?

For me, it feels like all what this will end up doing is losing consistency in our components since most would still need useSelect while some would use that one. Is the shorthand syntax here worth it enough to introduce this inconsistency and force contributors to learn a new hook?

ellatrix commented:
I don't get the merged API. 🤔 So we're _not_ reducing the amount of the subscriptions and selector calls by subscribing to one store only? We still go for the hardcoded store key strings?

adamziel commented:
@youknowriad @ellatrix I think this is pivoting into a more general discussion about selectors and registry API. In the next hour I'll spin a new issue to fully describe my thinking and potentially centralize all the different threads we're pulling. It may or may not be resolved quickly - what do you think about renaming this one to `__experimentalUseRegistrySelectors` until then?

adamziel commented:
Actually I changed my mind, let's at least try to continue here. I'm still happy to create a new issue for that anytime if someone brings it up.

@youknowriad This work is very emergent but I'll do my best to answer all the questions. It will be indirect at first, but I promise I'll get there!

### Overview 

I imagine the data flow across components and selectors as a complex graph like this:

<img width="576" alt="Zrzut ekranu 2020-11-9 o 13 42 34" src="https://user-images.githubusercontent.com/205419/98542592-75fc3f80-2291-11eb-9e91-7bb3bec2660b.png">

Circles are selectors. Some are simple and rely on a single store. Other selectors depend on selectors from other stores which means they require registry access. Since it's hard or impossible to tell which selectors rely on which stores, the components must listen to the changes on the entire registry:

<img width="351" alt="Zrzut ekranu 2020-11-9 o 13 43 46" src="https://user-images.githubusercontent.com/205419/98542771-b8258100-2291-11eb-923c-23c5d2601713.png">

This is essentially a star topology with every change affecting every node in one way or another. Adding nodes and interactions scales non-linearly and the graph-based data flow makes following the flow harder than necessary.

Now, what is emerging from my recent discussions with @gziolo is that this entire model could be simplified like this:

<img width="617" alt="Zrzut ekranu 2020-11-9 o 13 45 42" src="https://user-images.githubusercontent.com/205419/98542868-df7c4e00-2291-11eb-8396-16551e3b327d.png">

In this model, there are no registry selectors (an idea explored [in another thread](https://github.com/WordPress/gutenberg/pull/26733#issuecomment-723922544)). Instead, components pull the data directly from the stores they are in interested in. There is no concept of dependency graph at all, only a shallow list of dependent stores.

Of course, registry selectors sometimes contain complex logic - where should that logic go? Well, there already exists a mechanism that makes it possible to achieve exactly the same thing as registry selectors: React hooks. 

### Rationale

As noted [in another thread](https://github.com/WordPress/gutenberg/pull/26733#issuecomment-723922544), registry selectors may be a solution to a problem we just don't have anymore in core. If that's true, then they are just an inferior alternative to react hooks:

* All registry selectors are public API.
* It's unclear when to use registry selectors and when to use react hooks.
* Registry selectors create a complex dependency graph that's hard to reason about. It is also a structure where circular dependencies are possible.
* The complexity of the data flow scales non-linearly.
* They make it hard/impossible to reliably subscribe to only those stores that are relevant - a dependent selector may change one day and break some components.

Also, as @nerrad mentioned above, this idea came up multiple times already:

>  However, I do like the idea of optionally being able to subscribe to a single store (this has come up numerous times).

### Where we are now

As you noticed, for now this is just a duplicate API. I am a big proponent of having a moderate amount of well-placed shorthands, I believe they make the final code more succinct and easier to read. The maintenance and learning cost doesn't _seem_ large to me as it is just a simple wrapper fulfilling the mythical 100% code coverage goal. However, I am fairly new to Gutenberg so I will fully trust the judgement of more experienced folks. 

But let's look beyond that. Having a store-specific selector is a necessary first step for store-based subscriptions which I believe is an ideal state worth pursuing.

Besides the rationale laid out in the previous sections, there are potential performance gains involved. They are hard to assess upfront, for example I tried here: https://github.com/WordPress/gutenberg/pull/26692#issuecomment-722615094

I found that selectors-related code takes 10% of the blocking time when typing heavily. That's not 90% or 40% and there are definitely heavier tasks such as multiple re-renders going on. Still, it's 10% that we could potentially take down to 5% or 2%. I don't have a good idea how to reliably measure the impact of a full refactor. I tried to hack together a performance test, but it didn't show anything meaningful so I just abandoned this line of discussion until I'm able to get a better measurement.

### Surface area

None of what you'll see below includes the numbers for registry selectors - that would be much harder to analyze. Still, I took a few naive measures to assess how many components relies on a single store vs multiple stores. According to the simple command below:

```
$ git grep -A 1 'select(' | grep 'core/'  | grep "/components/" | gsed -E "s#^([^\t ]+)[\t ]([^']+)'([^']+)'.+#\1 \3#g" | uniq | awk '/ /{ print $1 $2 }' | uniq | awk -F: '{arr[$1]++}END{for (a in arr) print arr[a]": "a}' | sort -n
1 packages/block-directory/src/components/auto-block-uninstaller/index.js
1 packages/block-directory/src/components/downloadable-block-notice/index.js
...
$ !! |  awk -F: '{arr[$1]++}END{for (a in arr) print arr[a], a}' 
177 1
30 2
6 3
2 5
1 6
```

So there are 177 components relying on just a single store, 30 relying on two stores, and only a single component relying on 6 stores. 

I also measured how many components rely on each store:
```
$ git grep -A 1 'select(' | grep 'core/'  | grep "/components/" | gsed -E "s#^([^\t ]+)[\t ]([^']+)'([^']+)'.+#\1 \3#g" | uniq | awk '{print $2": "$1}' | sed "s/'//g" | uniq | sort
core/block-directory: packages/block-directory/src/components/auto-block-uninstaller/index.js:
core/block-directory: packages/block-directory/src/components/downloadable-block-list-item/index.js-
...
$ !! | awk -F: '{arr[$1]++}END{for (a in arr) print arr[a]": "a}' | sort -n
1: core/edit-navigation
1: core/edit-post/toggle-block-navigation
1: core/nux
1: core/viewport
3: core/data
4: core/block-directory
4: core/notices
5: core/edit-widgets
7: core/interface
9: core/keyboard-shortcuts
12: core/edit-site
18: core/blocks
53: core/edit-post
91: core/editor
131: core/block-editor
```

### Direct answers

I talked about the maintenance/learning cost earlier, but there was also a concern @youknowriad raised about the consistency:

> Is the shorthand syntax here worth it enough to introduce this inconsistency and force contributors to learn a new hook?

My thinking is that if we can get rid of the dependency graph, it would be worth it. But yes, at the moment this is just a tiny wrapper that doesn't do much.

> I don't get the merged API. 🤔 So we're not reducing the amount of the subscriptions and selector calls by subscribing to one store only?

I'd like to get there in the end, this is just a "foot in the door" so to speak. I'm happy to make it an experimental API for now.

> We still go for the hardcoded store key strings?

I'm all for object-based store keys, it's just beyond the scope of this PR. @gziolo is working on that change in another PR I believe - I don't want to steal his thunder :-)

### Other considerations

`useSelect` combined with eslint rule requiring unique variable names involves some awkward naming conventions with `_` prefixes, I really dislike that:

```js
const { parentBlockType, firstParentClientId, shouldHide } = useSelect(
		( select ) => {
			const {
				getBlockName,
				getBlockParents,
				getSelectedBlockClientId,
			} = select( 'core/block-editor' );
			const { hasBlockSupport } = select( 'core/blocks' );
			const selectedBlockClientId = getSelectedBlockClientId();
			const parents = getBlockParents( selectedBlockClientId );
			const _firstParentClientId = parents[ parents.length - 1 ];
			const parentBlockName = getBlockName( _firstParentClientId );
			const _parentBlockType = getBlockType( parentBlockName );
```

youknowriad commented:
> This is essentially a star topology with every change affecting every node in one way or another. Adding nodes and interactions scales non-linearly and following the graph-based data flow is harder than necessary.

How is this any different from any redux store, it's still a global store with everything in it. The only difference in the data module is that we try to split some smaller stores in an organized fashion low level stores but in the end, the top level store say edit-post or editor, still need too access data across stores and merge them in a single selector. If we don't do it at the selector level like we do with `registry selectors`, we'll end up duplicating this logic in components in `useSelect` calls. 

`getEditedPostContent` is a good example here, How would you write this with stores that are completely independent?

--- 

Overall, my opinion is that we're trying to solve the wrong problem here tbh. What are we trying to achieve with these? It'sn to clear to me, it's it about performance? Show me the numbers this has, I don't believe it has any meaningful impact personally but happy to be wrong here.

For me, until we know what we're trying to solve, it's very premature to introduce APIs we might regret later.

adamziel commented:
@youknowriad let's separate two threads we're touching on here.

### Thread 1 - this PR

> Is the shorthand syntax here worth it enough to introduce this inconsistency and force contributors to learn a new hook?

For now I don't have much more than my strong opinion about shorthands 😄 I hear what you're saying, let's revert this PR and revisit if / when the part below reaches a conclusion.

### Thread 2 - the bigger picture

> How is this any different from any redux store, it's still a global store with everything in it.

The big picture isn't inherently different, it's just this is a highly interactive app with a high number of state actions and subscriptions. It differs in details - the store is divided into smaller modules. IMHO that division could involve less code and a simpler mental model.

> What are we trying to achieve with these?

For me this part is more about the graphs and circles I posted above. I thing core data is pretty complex and I think we are very much capable of reducing that complexity. Performance gains are possible but since it's hard to take reliable measures upfront, that's more of a gamble than a solid basis.

> If we don't do it at the selector level like we do with registry selectors, we'll end up duplicating this logic in components in useSelect calls.

Not necessarily, my thinking is that these are interchangeable:

```js
// component.js:
	const postContent = useSelect(
		( select ) => select( 'core/editor' ).getEditedPostContent(),
		[]
	);

// selectors.js:
export const getEditedPostContent = createRegistrySelector(
	( select ) => ( state ) => {
		const postId = getCurrentPostId( state );
		const postType = getCurrentPostType( state );
		const record = select( 'core' ).getEditedEntityRecord(
			'postType',
			postType,
			postId
		);
		if ( record ) {
			if ( typeof record.content === 'function' ) {
				return record.content( record );
			} else if ( record.blocks ) {
				return serializeBlocks( record.blocks );
			} else if ( record.content ) {
				return record.content;
			}
		}
		return '';
	}
);
```

and

```js
// component.js:
	const postContent = useEditedPostContent();

// hooks.js
export const useEditedPostContent = () => {
	const [ postId, postType ] = useStoreSelectors( 'core/editor', ( { getCurrentPostId, getCurrentPostType } ) => ([
		getCurrentPostId(), getCurrentPostType()
	]));
	const record = useStoreSelectors( 'core', ( { getEditedEntityRecord }) => getEditedEntityRecord(
		'postType',
		postType,
		postId
	), [ postType, postId ]);
	if ( record ) {
		if ( typeof record.content === 'function' ) {
			return record.content( record );
		} else if ( record.blocks ) {
			return serializeBlocks( record.blocks );
		} else if ( record.content ) {
			return record.content;
		}
	}
	return '';
} );
```

Although I'm not yet sure what to do with usages like this:

https://github.com/WordPress/gutenberg/blob/1d6b8dcadcadcf82cce450824aecdbe55e7c7f45/packages/editor/src/store/actions.js#L232

adamziel commented:
Either way, the conclusion for now is to revert this PR. If I can come up with a solid design improvement for selectors, I'll create a new issue.

gziolo commented:
@adamziel, great job doing all explorations, I personally learned a ton. The discussion triggered in this PR following the merge is a clear sign that the data layer became very complex over time because of many legacy implications like the limitation HOCs created. Hooks open new possibilities that are worth exploring but the number of backward compatibility considerations makes the process extremely challenging. @youknowriad thanks for sharing more context which isn't obvious to other contributors that can't follow closely development happening in this area.
